File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/core/tea/TeaRepositoryTest.java
package coolpharaoh.tee.speicher.tea.timer.core.tea;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.app.Application;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;

import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase;

@ExtendWith(MockitoExtension.class)
class TeaRepositoryTest {
    @Mock
    TeaMemoryDatabase teaMemoryDatabase;
    @Mock
    TeaDao teaDao;

    private TeaRepository teaRepository;

    @BeforeEach
    void setUp() {
        TeaMemoryDatabase.setMockedDatabase(teaMemoryDatabase);
        when(teaMemoryDatabase.getTeaDao()).thenReturn(teaDao);

        teaRepository = new TeaRepository(new Application());
    }

    @Test
    void insertTea() {
        final Tea tea = new Tea();

        teaRepository.insertTea(tea);

        verify(teaDao).insert(tea);
    }

    @Test
    void updateTea() {
        final Tea tea = new Tea();

        teaRepository.updateTea(tea);

        verify(teaDao).update(tea);
    }

    @Test
    void deleteTeaById() {
        final long teaId = 1;

        teaRepository.deleteTeaById(teaId);

        verify(teaDao).deleteTeaById(teaId);
    }

    @Test
    void deleteAllTeas() {
        teaRepository.deleteAllTeas();

        verify(teaDao).deleteAll();
    }

    @Test
    void getTeas() {
        when(teaDao.getTeas()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeas();

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasOrderByActivity() {
        when(teaDao.getTeasOrderByActivity()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByActivity(false);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasInStockOrderByActivity() {
        when(teaDao.getTeasInStockOrderByActivity()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByActivity(true);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasOrderByAlphabetic() {
        when(teaDao.getTeasOrderByAlphabetic()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByAlphabetic(false);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasInStockOrderByAlphabetic() {
        when(teaDao.getTeasInStockOrderByAlphabetic()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByAlphabetic(true);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasOrderByVariety() {
        when(teaDao.getTeasOrderByVariety()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByVariety(false);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasStockOrderByVariety() {
        when(teaDao.getTeasInStockOrderByVariety()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByVariety(true);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasOrderByRating() {
        when(teaDao.getTeasOrderByRating()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByRating(false);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeasInStockOrderByRating() {
        when(teaDao.getTeasInStockOrderByRating()).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasOrderByRating(true);

        assertThat(teas).hasSize(2);
    }

    @Test
    void getTeaById() {
        final long teaId = 1;
        final Tea tea = new Tea();
        when(teaDao.getTeaById(teaId)).thenReturn(tea);

        final Tea teaById = teaRepository.getTeaById(teaId);

        assertThat(teaById).isEqualTo(tea);
    }

    @Test
    void getRandomTeaInStock() {
        final Tea tea = new Tea();
        when(teaDao.getRandomTeaInStock()).thenReturn(tea);

        final Tea randomTea = teaRepository.getRandomTeaInStock();

        assertThat(randomTea).isEqualTo(tea);
    }

    @Test
    void getTeasBySearchString() {
        final String searchString = "search";
        when(teaDao.getTeasBySearchString(searchString)).thenReturn(Arrays.asList(new Tea(), new Tea()));

        final List<Tea> teas = teaRepository.getTeasBySearchString(searchString);

        assertThat(teas).hasSize(2);
    }
}

File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/core/tea/TeaRepositoryTest.kt
package coolpharaoh.tee.speicher.tea.timer.core.tea

import android.app.Application
import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase
import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase.Companion.setMockedDatabase
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class TeaRepositoryTest {
    @MockK
    lateinit var teaMemoryDatabase: TeaMemoryDatabase
    @RelaxedMockK
    lateinit var teaDao: TeaDao

    private var teaRepository: TeaRepository? = null

    @BeforeEach
    fun setUp() {
        setMockedDatabase(teaMemoryDatabase)
        every { teaMemoryDatabase.teaDao } returns teaDao

        teaRepository = TeaRepository(Application())
    }

    @Test
    fun insertTea() {
        val tea = Tea()

        teaRepository!!.insertTea(tea)

        verify { teaDao.insert(tea) }
    }

    @Test
    fun updateTea() {
        val tea = Tea()

        teaRepository!!.updateTea(tea)

        verify { teaDao.update(tea) }
    }

    @Test
    fun deleteTeaById() {
        val teaId: Long = 1

        teaRepository!!.deleteTeaById(teaId)

        verify { teaDao.deleteTeaById(teaId) }
    }

    @Test
    fun deleteAllTeas() {
        teaRepository!!.deleteAllTeas()

        verify { teaDao.deleteAll() }
    }

    @Test
    fun getTeas() {
        every { teaDao.getTeas() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.teas

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasOrderByActivity() {
        every { teaDao.getTeasOrderByActivity() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByActivity(false)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasInStockOrderByActivity() {
        every { teaDao.getTeasInStockOrderByActivity() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByActivity(true)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasOrderByAlphabetic() {
        every { teaDao.getTeasOrderByAlphabetic() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByAlphabetic(false)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasInStockOrderByAlphabetic() {
        every { teaDao.getTeasInStockOrderByAlphabetic() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByAlphabetic(true)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasOrderByVariety() {
        every { teaDao.getTeasOrderByVariety() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByVariety(false)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasStockOrderByVariety() {
        every { teaDao.getTeasInStockOrderByVariety() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByVariety(true)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasOrderByRating() {
        every { teaDao.getTeasOrderByRating() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByRating(false)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeasInStockOrderByRating() {
        every { teaDao.getTeasInStockOrderByRating() } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasOrderByRating(true)

        assertThat(teas).hasSize(2)
    }

    @Test
    fun getTeaById() {
        val teaId: Long = 1
        val tea = Tea()
        every { teaDao.getTeaById(teaId) } returns tea

        val teaById = teaRepository!!.getTeaById(teaId)

        assertThat(teaById).isEqualTo(tea)
    }

    @Test
    fun getRandomTeaInStock() {
        val tea = Tea()
        every { teaDao.getRandomTeaInStock() } returns tea

        val randomTea = teaRepository!!.randomTeaInStock

        assertThat(randomTea).isEqualTo(tea)
    }

    @Test
    fun getTeasBySearchString() {
        val searchString = "search"
        every { teaDao.getTeasBySearchString(searchString) } returns listOf(Tea(), Tea())

        val teas = teaRepository!!.getTeasBySearchString(searchString)

        assertThat(teas).hasSize(2)
    }
}